### PR TITLE
[FLINK-32317] Enrich metadata in CR error field

### PR DESCRIPTION
**Pull request by darenwkt**
## What is the purpose of the change

CR Error field is improved in https://issues.apache.org/jira/browse/FLINK-29708.

The error field is more structured with exception type, stackTrace, additionalMetadata, etc.

 

This ticket is a proposal to expose a config ("kubernetes.operator.exception.metadata.mapper") to enrich the additionalMetadata further.

 

The config consists of key-value pairs, for example:

```
kubernetes.operator.exception.metadata.mapper: IOException:Found IOException,403:Found 403 error code
```

The key is a REGEX string that will be used to match against the whole stack trace and if found, the value will be added to additionalMetadata. For example:

```
apiVersion: flink.apache.org/v1beta1
kind: FlinkSessionJob
....
  name: basic-session-job-example
  namespace: default
  resourceVersion: "70206149"
  uid: 916ea8f5-0821-4839-9953-2db9678c3fc9
spec:
  deploymentName: basic-session-deployment-example
  job:
    args: []
    jarURI: https://test-s3.s3.amazonaws.com/doubleExecute.jar
    parallelism: 4
    state: running
    upgradeMode: stateless
status:
  error: '{"type":"org.apache.flink.kubernetes.operator.exception.ReconciliationException","message":"java.io.IOException:
    Server returned HTTP response code: 403 for URL: https://test-s3.s3.amazonaws.com/doubleExecute.jar","additionalMetadata":{"exceptionMapper":["Found
    403 error code","Found IOException"]},"throwableList":[{"type":"java.io.IOException","message":"Server
    returned HTTP response code: 403 for URL: https://test-s3.s3.amazonaws.com/doubleExecute.jar","additionalMetadata":{"exceptionMapper":["Found
    403 error code"]}}]}'
...
```
 

## Brief change log

- Added new operator config for user to specify REGEX to extract info from exception and enrich CR error field with it.

## Verifying this change

- Added unit test
- Sanity test by building image and testing in minikube

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: (no)
  - Core observer or reconciler logic that is regularly executed: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

**Comment by gyfora** [comment]
I wonder if label mapper would be more appropriate naming for the feature/config/status field:

```
additionalMetadata":{"labels":["x", "y"]}...
```

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java [hunk 1]
```diff
@@ -244,6 +245,14 @@ public static String operatorConfigKey(String key) {
                     .withDescription(
                             "Maximum number of throwable to be included in CR status error field.");
 
+    @Documentation.Section(SECTION_DYNAMIC)
+    public static final ConfigOption<Map<String, String>> OPERATOR_EXCEPTION_LABEL_MAPPER =
+            operatorConfig("exception.metadata.mapper")
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:250]: The config option should also be renamed to `exception.label.mapper` and then please regenerate the docs.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java [hunk 2]
```diff
@@ -244,6 +245,14 @@ public static String operatorConfigKey(String key) {
...
+                    .mapType()
+                    .defaultValue(new HashMap<>())
+                    .withDescription(
+                            "Key-Value pair where key is the REGEX to filter through the exception messages and value is the string to be included in CR status error metadata field if the REGEX matches. Expected format: headerKey1:headerValue1,headerKey2:headerValue2.");
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:254]: Please update the description to refer to label instead of metadata

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/FlinkResourceExceptionUtils.java [hunk 3]
```diff
@@ -29,14 +29,18 @@
...
 import java.util.Map;
 import java.util.Optional;
+import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
 /** Flink Resource Exception utilities. */
 public final class FlinkResourceExceptionUtils {
 
     private static final ObjectMapper objectMapper = new ObjectMapper();
+    public static final String LABEL_MAPPER_IDENTIFIER = "labels";
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/FlinkResourceExceptionUtils.java:43]: Could we simply call this constant `LABELS` or `LABELS_KEY?`

File: helm/flink-kubernetes-operator/conf/flink-conf.yaml [hunk 4]
```diff
@@ -43,6 +43,7 @@ parallelism.default: 1
 # kubernetes.operator.exception.stacktrace.max.length: 2048
 # kubernetes.operator.exception.field.max.length: 2048
 # kubernetes.operator.exception.throwable.list.max.count: 2
+# kubernetes.operator.exception.metadata.mapper: Job has already been submitted:duplicatedJobFound,Server returned HTTP response code:httpResponseCodeFound
```
Comment (gyfora) [inline helm/flink-kubernetes-operator/conf/flink-conf.yaml:46]: Please update this as well after renaming the config key

**Review by gyfora — CHANGES_REQUESTED** [review]

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/FlinkResourceExceptionUtils.java [hunk 5]
```diff
@@ -129,26 +144,56 @@ public static Optional<String> getSubstringWithMaxLength(String str, int limit)
...
         }
 
+        labelMapper
+                .entrySet()
+                .forEach(
+                        (entry) -> {
+                            Pattern pattern = Pattern.compile(entry.getKey());
+
+                            org.apache.flink.util.ExceptionUtils.findThrowable(
+                                            throwable, t -> pattern.matcher(t.getMessage()).find())
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/FlinkResourceExceptionUtils.java:178]: Can you please add a test for Exceptions with null message? (such as a nullpointer) I think this code would fail

**Review by gyfora — CHANGES_REQUESTED** [review]

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/FlinkResourceExceptionUtils.java (hunk 5 above)
Comment (darenwkt) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/FlinkResourceExceptionUtils.java]: Hi Gyula, thanks for catching this, I have fixed the code and added the test, thank you

**Comment by gyfora** [comment]
you need ro run `mvn spotless:apply`

**Review by gyfora — APPROVED** [review]
